channeldb/payment_control.go

```diff
@@ -183,6 +214,68 @@ func (p *PaymentControl) InitPayment(paymentHash lntypes.Hash,
 	return updateErr
 }
 
+// createPaymentIndexEntry adds a payment index keyed by sequence number with
+// the payment hash as its value.
+func createPaymentIndexEntry(tx kvdb.RwTx, sequenceNumber []byte,
+	hash lntypes.Hash) error {
+
+	indexes := tx.ReadWriteBucket(paymentsIndexBucket)
+	if indexes == nil {
+		return errNoIndexBucket
+	}
+
+	return indexes.Put(sequenceNumber, hash[:])
+}
+
+// fetchPaymentIndexEntry gets the payment hash for the sequence number provided
+// from our payment indexes bucket.
+func (p *PaymentControl) fetchPaymentIndexEntry(sequenceNumber uint64) (
+	*lntypes.Hash, error) {
+
+	var hash lntypes.Hash
+
+	if err := kvdb.View(p.db, func(tx walletdb.ReadTx) error {
+		indexBucket := tx.ReadBucket(paymentsIndexBucket)
+		if indexBucket == nil {
+			return errNoIndexBucket
+		}
+
+		key := make([]byte, 8)
+		byteOrder.PutUint64(key, sequenceNumber)
+
+		indexValue := indexBucket.Get(key)
+		if indexValue == nil {
+			return errNoIndex
+		}
+
+		r := bytes.NewReader(indexValue)
+
+		var err error
+		hash, err = deserializePaymentIndex(r)
+		return err
+
+	}); err != nil {
+		return nil, err
+	}
+
+	return &hash, nil
+}
+
+// deserializePaymentIndex deserializes a payment index entry.
+func deserializePaymentIndex(r io.Reader) (lntypes.Hash, error) {
+	var paymentHash [lntypes.HashSize]byte
+	if err := ReadElements(r, &paymentHash); err != nil {
+		return lntypes.Hash{}, err
+	}
+
+	hash, err := lntypes.MakeHash(paymentHash[:])
```

Comment:
good point, that is still tbd as to the best way to do that. more than likely we will start to index payments by payment address since that will be a uniform identifier that also works for MPP. when you start getting to recurring payments we would need to index by set_id in order to differentiate logical payments to the same invoice. 

this could be implemented by adding a type byte to the index to signify which index the following 32-byte correspond to, but i don't know how much effort it would be for this PR. perhaps the best approach would be to deal with these details when we get there, and say that `len(value) == 32` means payment hash. when we add the newer indexes, they could receive this type byte and be 33 bytes. i wouldn't want to block progress on this for some undecided details on AMP, but all in all i think we should be able to extend it as we need 👍